Assets/Mirage/Runtime/InterestManagement/InterestManager.cs

```diff
@@ -0,0 +1,286 @@
+        /// <param name="skip">a player who should not receive the message</param>
+        /// <returns>Total amounts of messages sent</returns>
+        protected virtual int Send(NetworkIdentity identity, ArraySegment<byte> data, int channelId = Channel.Reliable, INetworkPlayer skip = null)
+        {
+            int count = 0;
+
+            foreach (INetworkPlayer player in Observers(identity))
+            {
+                if (player == null) continue;
+
+                if (player != skip)
+                {
+                    // send to all connections, but don't wait for them
+                    player.Send(data, channelId);
+                    count++;
+                }
+            }
+
+            return count;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void CheckGrowVisibilitySystem()
+        {
+            var stopWatch = Stopwatch.StartNew();
+
+            if (_visibilitySystems == null)
+            {
+                _visibilitySystems = new ObserverData[_initialSystems];
+            }
+            else if (_visibilitySystems[_visibilitySystems.Length - 1].System == default)
+            {
+                // We need to copy and expand our visibility system.
+                if (logger.logEnabled)
+                    logger.Log("[Interest Manager] - Visibility system is expanding array. If this is happening often. Please set initial systems in constructor.");
+
+                var newData = new ObserverData[_initialSystems + _initialSystems];
+
+                _visibilitySystems.CopyTo(newData, 0);
+
+                _visibilitySystems = newData;
+            }
+
+            stopWatch.Stop();
+
+            if (logger.logEnabled)
+                logger.Log($"[Interest Manager] - CheckGrowVisibilitySystem Method Execution Time: {stopWatch.Elapsed.TotalMilliseconds} ms");
+        }
+
+        /// <summary>
+        ///     Register a specific interest management system to the interest manager.
+        /// </summary>
+        /// <param name="system">The system we want to register in the interest manager.</param>
+        internal void RegisterVisibilitySystem(ref ObserverData system)
+        {
+            CheckGrowVisibilitySystem();
+
+            var stopWatch = Stopwatch.StartNew();
+
+            if (_visibilitySystems.Equals(system))
+            {
+                logger.LogWarning(
+                    "[InterestManager] - System already register to interest manager. Please check if this was correct.");
+
+                return;
+            }
+
+            if(logger.logEnabled)
+                logger.Log($"[Interest Manager] - Registering system {system} to our manager.");
+
+            for (int i = 0; i < _visibilitySystems.Length; i++)
+            {
+                if(!_visibilitySystems[i].Equals(default)) continue;
+
+                _visibilitySystems[i] = system;
+
+                stopWatch.Stop();
```

Comment:
Stopping watch here means it'll only measure the first iteration of the loop

Reply:
That's correct it is suppose to do that. If you notice below after the log I return out of the method. So stopping there needs to be done. I have to loop through the array to get the info I need to check if its already registered.  If it finds an open spot then it attaches to that index spot and leaves

Reply:
maybe a `break;` would be better here so that you dont have the duplicate stop/log/return blocks

Reply:
Sounds good can do.